primitives/api/proc-macro/src/impl_runtime_apis.rs

```diff
@@ -383,9 +387,9 @@ fn generate_runtime_api_base_structures() -> Result<TokenStream> {
 			fn commit_on_ok<R, E>(&self, res: &std::result::Result<R, E>) {
 				if *self.commit_on_success.borrow() {
 					if res.is_err() {
-						self.changes.borrow_mut().discard_prospective();
+						self.changes.borrow_mut().rollback_transaction();
```

Comment:
There is no longer something that is called *prospective*. Only values existing in transactions and values that are already committed and cannot be rolled back. I think speaking about *prospective* is confusing at this point.

Runtime is responsible for balancing its transactions. Failing to do so is a hard programming error and not a runtime error that we need to handle. An easily avoidable error, too: Runtime will use the `with_transaction` wrapper which guarantees balanced transactions. That said I am strictly against complicating the code in order to cover up programming errors. It will be actively harmful to not panic on such an error.

Reply:
Just pushed the changes that protect the client from unbalanced transaction in the runtime.